### PR TITLE
[Feat] Initial Implementation of PodAutoscaler Reconciler

### PR DESCRIPTION
**Pull request by kr11**
## Pull Request Description

- Controller and Reconciler: Implemented a PodAutoscaler controller and reconciler that automatically generates an HPA upon deployment of an AIBrix PodAutoscaler.
- Demo YAMLs: Included to demonstrate functionality.
- Types Adjustments: Updated `api/autoscaling/v1alpha1/podautoscaler_types.go`.
- Documentation: Added README for setup and usage guidance.

## Related Issues
Resolves: #24 

**Important: Before submitting, please complete the description above and review the checklist below.**

---

<details>
<summary><strong>Contribution Guidelines (Expand for Details)</strong></summary>

<p>We appreciate your contribution to aibrix! To ensure a smooth review process and maintain high code quality, please adhere to the following guidelines:</p>

<h3>Pull Request Title Format</h3>
<p>Your PR title should start with one of these prefixes to indicate the nature of the change:</p>
<ul>
    <li><code>[Bug]</code>: Corrections to existing functionality</li>
    <li><code>[CI]</code>: Changes to build process or CI pipeline</li>
    <li><code>[Docs]</code>: Updates or additions to documentation</li>
    <li><code>[API]</code>: Modifications to aibrix's API or interface</li>
    <li><code>[CLI]</code>: Changes or additions to the Command Line Interface</li>
    <li><code>[Misc]</code>: For changes not covered above (use sparingly)</li>
</ul>
<p><em>Note: For changes spanning multiple categories, use multiple prefixes in order of importance.</em></p>

<h3>Submission Checklist</h3>
<ul>
    <li>[ ] PR title includes appropriate prefix(es)</li>
    <li>[ ] Changes are clearly explained in the PR description</li>
    <li>[ ] New and existing tests pass successfully</li>
    <li>[ ] Code adheres to project style and best practices</li>
    <li>[ ] Documentation updated to reflect changes (if applicable)</li>
    <li>[ ] Thorough testing completed, no regressions introduced</li>
</ul>

<p>By submitting this PR, you confirm that you've read these guidelines and your changes align with the project's contribution standards.</p>

</details>

### CODE REVIEW

File: api/autoscaling/v1alpha1/podautoscaler_types.go [hunk 1]
```diff
@@ -17,37 +17,91 @@ limitations under the License.
...
-	Foo string `json:"foo,omitempty"`
+	TargetValue string `json:"targetValue"`
+
+	// MetricsSources defines a list of sources from which metrics are collected to make scaling decisions.
+	MetricsSources []MetricSource `json:"metricsSources,omitempty"`
+
+	// ScalingStrategy defines the strategy to use for scaling.
+	// 1. "HPA" for using Kubernetes HPA
+	// 2. "Custom" for a custom scaling mechanism.
+	ScalingStrategy string `json:"scalingStrategy"`
```
Comment (Jeffwan) [inline api/autoscaling/v1alpha1/podautoscaler_types.go]: This should be pre-configured. Let's create a type for `ScalingStrategy`. do not use string here. here's the example https://github.com/aibrix/aibrix/blob/d3bd97115f7791af6dc38f4af610ca7b123c76f0/api/model/v1alpha1/modeladapter_types.go#L89

File: api/autoscaling/v1alpha1/podautoscaler_types.go [hunk 2]
```diff
@@ -17,37 +17,91 @@ limitations under the License.
...
+
+	// AdditionalConfig provides a place for custom settings that are not defined as standard fields.
+	AdditionalConfig map[string]string `json:"additionalConfig,omitempty"`
```
Comment (Jeffwan) [inline api/autoscaling/v1alpha1/podautoscaler_types.go]: we probably do not need it in autoscaling. If this is not referenced in your code, let's remove it. 

HPA need more policies like cooldown, step duration, stabilization etc

File: api/autoscaling/v1alpha1/podautoscaler_types.go [hunk 3]
```diff
@@ -17,37 +17,91 @@ limitations under the License.
...
+// MetricSource defines an endpoint and path from which metrics are collected.
+type MetricSource struct {
+	// e.g. service1.example.com
+	Endpoint string `json:"endpoint"`
+	// e.g. /api/metrics/cpu
+	Path string `json:"path"`
 }
 
 // PodAutoscalerStatus defines the observed state of PodAutoscaler
+// PodAutoscalerStatus reflects the current state of the PodAutoscaler,
```
Comment (Jeffwan) [inline api/autoscaling/v1alpha1/podautoscaler_types.go]: line 85 vs 86. clean them up?

File: api/autoscaling/v1alpha1/podautoscaler_types.go [hunk 4]
```diff
@@ -17,37 +17,91 @@ limitations under the License.
...
+// including the current number of replicas, operational status, and other metrics.
 type PodAutoscalerStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
-}
+	// ServiceName is the name of the Kubernetes Service that routes traffic to the target, scaled by this PodAutoscaler.
```
Comment (Jeffwan) [inline api/autoscaling/v1alpha1/podautoscaler_types.go]: do we have service name?

File: config/samples/autoscaling_v1alpha1_podautoscaler.yaml [hunk 5]
```diff
@@ -1,9 +1,18 @@
...
-  # TODO(user): Add fields here
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: nginx-deployment
+  minReplicas: 1
+  maxReplicas: 10
+  targetMetric: "CPU"
+  targetValue: "10"
+  scalingStrategy: "HPA"
```
Comment (Jeffwan) [inline config/samples/autoscaling_v1alpha1_podautoscaler.yaml:18]: append an empty line at the end

File: pkg/controller/podautoscaler/README.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
+# AIBrix PodAutoscaler Demo
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/README.md]: normally we do not put README.md under controller folder. Please create `docs/development` folder and move this doc to there.

File: pkg/controller/podautoscaler/hpa_resources.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+import (
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+)
+
+var (
+	controllerKind = pa_v1.GroupVersion.WithKind("PodAutoScaler") // Define the resource type for the controller
+)
+
+func getHPANameFromPa(pa *pa_v1.PodAutoscaler) string {
+	return pa.Name + "-hpa"
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/hpa_resources.go]: minor: in golang, it's better to use `fmt.Sprintf`

File: pkg/controller/podautoscaler/hpa_resources.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// MakeHPA creates an HPA resource from a PodAutoscaler resource.
+func MakeHPA(pa *pa_v1.PodAutoscaler, ctx context.Context) *autoscalingv2.HorizontalPodAutoscaler {
+	_log := log.FromContext(ctx)
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/hpa_resources.go]: let's use `klog` instead for consistency

File: pkg/controller/podautoscaler/podautoscaler_controller.go [hunk 9]
```diff
@@ -18,14 +18,19 @@ package podautoscaler
 
 import (
 	"context"
+	pav1alpha1 "github.com/aibrix/aibrix/api/autoscaling/v1alpha1"
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/podautoscaler_controller.go]: let's rename to `autoscalingv1alpha1` instead

File: pkg/controller/podautoscaler/podautoscaler_controller.go [hunk 10]
```diff
@@ -77,19 +78,62 @@ type PodAutoscalerReconciler struct {
...
-	// TODO(user): your logic here
+	// Generate a corresponding HorizontalPodAutoscaler
+	hpa := MakeHPA(&pa, ctx)
+	hpaName := types.NamespacedName{
+		Name:      hpa.Name,
+		Namespace: hpa.Namespace,
+	}
+
+	var existingHPA autoscalingv2.HorizontalPodAutoscaler
+	err := r.Get(ctx, hpaName, &existingHPA)
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/podautoscaler_controller.go:112]: did you run into issue here? you probably need to write in this way? `existingHPA := &autoscalingv2.HorizontalPodAutoscaler{}`?

File: pkg/controller/podautoscaler/podautoscaler_controller.go [hunk 11]
```diff
@@ -77,19 +78,62 @@ type PodAutoscalerReconciler struct {
...
+// HorizontalPodAutoscalers based on the PodAutoscaler specifications.
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
 func (r *PodAutoscalerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	// Implement a timeout for the reconciliation process.
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	log_ := log.FromContext(ctx)
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/podautoscaler_controller.go]: let's use `klog` for consistency

File: pkg/controller/podautoscaler/podautoscaler_controller.go [hunk 12]
```diff
@@ -77,19 +78,62 @@ type PodAutoscalerReconciler struct {
...
+			return ctrl.Result{}, err
+		}
+	} else if err != nil {
+		// Error occurred while fetching the existing HPA, report the error and requeue.
+		log_.Error(err, "Failed to get HPA")
+		return ctrl.Result{}, err
+	} else {
+		// Update the existing HPA if it already exists.
+		log_.Info("Updating existing HPA", "HPA.Namespace", existingHPA.Namespace, "HPA.Name", existingHPA.Name)
+		hpa.ResourceVersion = existingHPA.ResourceVersion
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/podautoscaler_controller.go]: why do you update the revision here?

Comment (Jeffwan) [reply pkg/controller/podautoscaler/podautoscaler_controller.go]: Should you check the replica or other settings?

File: .gitignore [hunk 13]
```diff
@@ -20,3 +20,5 @@
 # Go workspace file
 go.work
 go.work.sum
+.idea
```
Comment (Jeffwan) [inline .gitignore]: this is great. however, liguang picks this up. let's leave this minor change to him as onboarding task.

Comment (kr11) [reply .gitignore]: Sure, the changes have been reverted

File: api/autoscaling/v1alpha1/podautoscaler_types.go (hunk 1 above)
Comment (kr11) [reply api/autoscaling/v1alpha1/podautoscaler_types.go]: Done, I added a type `ScalingStrategyType`

File: api/autoscaling/v1alpha1/podautoscaler_types.go (hunk 2 above)
Comment (kr11) [reply api/autoscaling/v1alpha1/podautoscaler_types.go]: I'd like to think it over. Should we include all HPA parameters in the spec, which might make it too aligned with HPA, or maintain flexibility by placing them in the config or params?"

File: api/autoscaling/v1alpha1/podautoscaler_types.go (hunk 3 above)
Comment (kr11) [reply api/autoscaling/v1alpha1/podautoscaler_types.go]: Fixed

File: api/autoscaling/v1alpha1/podautoscaler_types.go (hunk 4 above)
Comment (kr11) [reply api/autoscaling/v1alpha1/podautoscaler_types.go]: Not yet. I have removed  the commented-out `ServiceName` and `MetricsServiceName`

File: config/samples/autoscaling_v1alpha1_podautoscaler.yaml (hunk 5 above)
Comment (kr11) [reply config/samples/autoscaling_v1alpha1_podautoscaler.yaml:18]: Removed the blank line in `autoscaling_v1alpha1_demo_nginx.yaml`

File: pkg/controller/podautoscaler/README.md (hunk 6 above)
Comment (kr11) [reply pkg/controller/podautoscaler/README.md]: Done. Moved to `docs/tutorial/podautoscaler/README.md`

File: pkg/controller/podautoscaler/hpa_resources.go (hunk 7 above)
Comment (kr11) [reply pkg/controller/podautoscaler/hpa_resources.go]: Done.👍

File: pkg/controller/podautoscaler/hpa_resources.go (hunk 8 above)
Comment (kr11) [reply pkg/controller/podautoscaler/hpa_resources.go]: changed to`klog.V(3).ErrorS` and `klog.V(3).InfoS`

File: pkg/controller/podautoscaler/podautoscaler_controller.go (hunk 9 above)
Comment (kr11) [reply pkg/controller/podautoscaler/podautoscaler_controller.go]: Sure. Done.

File: pkg/controller/podautoscaler/podautoscaler_controller.go (hunk 11 above)
Comment (kr11) [reply pkg/controller/podautoscaler/podautoscaler_controller.go]: Done. changed to `klog.V(3)`

File: pkg/controller/podautoscaler/podautoscaler_controller.go (hunk 10 above)
Comment (kr11) [reply pkg/controller/podautoscaler/podautoscaler_controller.go:112]: It is running normally for me 😂 , my env (go1.22.5). I have fixed the code as your style.

File: pkg/controller/podautoscaler/podautoscaler_controller.go (hunk 12 above)
Comment (kr11) [reply pkg/controller/podautoscaler/podautoscaler_controller.go]: Sure, I'll re-check the code logic

File: api/autoscaling/v1alpha1/podautoscaler_types.go (hunk 2 above)
Comment (Jeffwan) [reply api/autoscaling/v1alpha1/podautoscaler_types.go]: this is something we need to discuss. At this moment, we do not want to replicate everything. We'd better think of our own algorithms and then consider the common parameters and those can be supported earlier.

File: pkg/controller/podautoscaler/podautoscaler_controller.go (hunk 10 above)
Comment (Jeffwan) [reply pkg/controller/podautoscaler/podautoscaler_controller.go:112]: Got you. This is ok.

File: pkg/controller/podautoscaler/podautoscaler_controller.go (hunk 12 above)
Comment (Jeffwan) [reply pkg/controller/podautoscaler/podautoscaler_controller.go]: We can remove this line for short term. Make sure HPA object is up-to-date could be a separate PR.

File: api/autoscaling/v1alpha1/podautoscaler_types.go (hunk 2 above)
Comment (kr11) [reply api/autoscaling/v1alpha1/podautoscaler_types.go]: > this is something we need to discuss. At this moment, we do not want to replicate everything. We'd better think of our own algorithms and then consider the common parameters and those can be supported earlier.

AdditionalConfig has been removed temporarily

File: pkg/controller/podautoscaler/podautoscaler_controller.go (hunk 12 above)
Comment (kr11) [reply pkg/controller/podautoscaler/podautoscaler_controller.go]: @Jeffwan the line has been removed temporarily

**Pull request merged**